### ISSUE TITLE
Entities not loading or graphic artefact on release build

### ISSUE THREAD

**Issue by FalseIlyu** [openblack/openblack#395]
All information are gathered on a windows build of openblack with vs code through Visual Studio compiler and builder, Configuration is Visual Studio 2022 (vcpkg), screenshot are all from the release build preset.
- Openblack is on ac5c42f.
- vcpkg is on 440075a9.

Visual problems seems to appear only on release and are renderer dependent, with Direct3D12 displaying an empty map.
![Openblack_direct12](https://user-images.githubusercontent.com/83550517/164972754-cf047448-e53e-4988-a8f4-2a6f2533e3da.png)
Other renderers will also display "GPU artifacts" with the most extreme case being redering with OpenGLES where the view get clutered.
![Openblack_OpenGLES](https://user-images.githubusercontent.com/83550517/164972917-d676b2e1-bd24-456c-92dd-286469318bae.png)
Even in Direct3D12 mode artifact can appear, as the bug vary in intensity each launch.

In the log there is a good prevalence of this error message.
`[2022-04-24 13:00:00.642] [game] [error] [L3DMesh.cpp:81] Failed to open L3DSubMesh from file: buffer`
Which if the root is reading from a buffer in an inconsistent state is in line with the way severity will vary from rendering backend and launch.

Here are the log for Release
[openblack.log](https://github.com/openblack/openblack/files/8549501/openblack.log)
Here are the log for debug, where the module appear bug the error are still there.
[openblack.log](https://github.com/openblack/openblack/files/8549508/openblack.log)

This does not appear to be a code issue as my test to track the commit of apparition have led me to version of openblack were i know this problem wasn't there 543c2c987bbab0d7fa0f46baa2300821281b4e46 for example worked fine during my investigation of #382.

Edit : I tried deleting the local repo and cloning it again, the problem persist.

**Comment by FalseIlyu** [comment]
There might still be a problem of configuration on my end or with my system, but i have reason to think my first assessement of it not being a bug caused by change in the code was mistaken.
Reason : I tried to setup a breakpoint in L3DMesh.cpp:81 on 543c2c9 and it didn't trigger while it it on ac5c42af63ac8c218b036f62f28f344ffd4e4310.
The bug still happen on release for both though, but that means the appartion of `[2022-04-24 13:00:00.642] [game] [error] [L3DMesh.cpp:81] Failed to open L3DSubMesh from file: buffer` in debug build could be traced to a626509604b49fd5444a4cd4621ec0dc60b6b96a

**Comment by bwrsandman** [comment]
I've replicated the entities not being drawn on Windows in Release mode.
I am running a bisect but the bug exists in https://github.com/openblack/openblack/commit/543c2c987bbab0d7fa0f46baa2300821281b4e46 for me.

**Comment by FalseIlyu** [comment]
Do you mean the entities not being drawn in release build or the game hitting`2022-04-24 13:00:00.642] [game] [error] [L3DMesh.cpp:81] Failed to open L3DSubMesh from file: buffer` in debug.
Because the first one happens on all build i tried including https://github.com/openblack/openblack/commit/543c2c987bbab0d7fa0f46baa2300821281b4e46 but error started being shown in debug build only after https://github.com/openblack/openblack/commit/a626509604b49fd5444a4cd4621ec0dc60b6b96a, for the moment i test both here because they seemed linked to me but there may be two separate issue at hand.

**Comment by bwrsandman** [comment]
I mean the entities not showing up

**Comment by bwrsandman** [comment]
I can replicate on linux but only with the vcpkg preset. Could be tied to a bgfx version.

**Comment by bwrsandman** [comment]
I've replicated the issue all the way back to 8d0c820e016cc7d343ec345224b3c3c60569bb2b and it's probably even older...
I remember a similar bug a few years ago.

**Comment by bwrsandman** [comment]
Bug disappears when turning off terrain rendering. Turning off sky also affects it.

**Comment by FalseIlyu** [comment]
I assume all your test test are on linux, would it be possible to know the build preset ?

The following is still windows with Visual Studio vcpkg configuration.
On my end i tried to look at behavior and 8d0c820 seems to work fine, if the deformation are not very severe i could just not be aware of them. For the effect of turning on and off sky and/or terrain rendering on ac5c42a i couldn't get any effet (i tried with multiple backend renderer : OpenGL, OpenGLES, Direct3D12, Vulkan, same result) on the other end disabling water always give a result like the second screenshot i share.

**Comment by bwrsandman** [comment]
I tested on two linux computers and was only able to replicate on `ninja-multi-vcpkg-release` Vulkan and only on one of the machines.

**Comment by FalseIlyu** [comment]
Compiling the renderer with `#pragma optimize( "", off )` make the problem go away.
I had this idea looking for similar problem and found this https://github.com/libigl/libigl/issues/1669.

**Comment by FalseIlyu** [comment]
Making identity static solves the problem, i'm not super confident but if identity isn't used except as reference couldn't it be that the compiler optimize out its initialisation?

**Comment by bwrsandman** [comment]
I pushed a fix, can you validate?

**Comment by bwrsandman** [comment]
Looks like we converged on the right answer at the same time 😃 
It's not really the compiler that optimizes it out. It's just that the memory region becomes free and some other stack data probably comes in and overrides it.

**Comment by bwrsandman** [comment]
The memory for a stack local variable is only valid within that scope.

```
// a doesn't exist
// &a is an invalid region of memory with garbage in it
{
    int a = 42;
    // a exists
    // &a points to a and it guaranteed to be the only data the stack will assign there
}
// a doesn't exist
// &a is a free region of memory that any new stack can use to store other data
// all references to &a here is undefined behaviour in regards to the value of a

```

**Comment by bwrsandman** [comment]
```c++
if (mesh->IsBoned())
{
	submitDesc.modelMatrices = mesh->GetBoneMatrices().data(); // The scope of this matrix is valid as long as mesh is
	submitDesc.matrixCount = static_cast<uint8_t>(mesh->GetBoneMatrices().size());
	// TODO(bwrsandman): Get animation frame instead of default
}
else
{
	const auto identity = glm::mat4(1.0f); // the cope of identity exists only here
	submitDesc.modelMatrices = &identity;
	submitDesc.matrixCount = 1;
} // this is the end of the cope of identity
submitDesc.isSky = false;
submitDesc.skyType = desc.sky.GetCurrentSkyType();

// This function is called using submitDesc which has an invalid pointer to the old memory location of identity
// It is likely that DrawMesh's stack memory will overlap with the memory of &identity
DrawMesh(mesh, submitDesc, std::numeric_limits<uint8_t>::max());
```
